File: 240528/트리의 지름/diameter-of-tree.py
import sys

input = sys.stdin.readline

n = int(input())

G = [[] for _ in range(n)]
visited = [False for _ in range(n)]
dist = [0 for _ in range(n)]

for _ in range(n - 1):
    s, e, w = map(int, input().split())
    G[s - 1].append((e - 1, w))
    G[e - 1].append((s - 1, w))
    
order = 0

def dfs(x):
    global order

    for to_idx, to_dist in G[x]:
        if not visited[to_idx]:
            order = order + to_dist
            dist[to_idx] = order
            visited[to_idx] = True
            dfs(to_idx)
            order = order - to_dist

visited[0] = True
dfs(0)
max_val = 0
max_idx = 0
for i, value in enumerate(dist):
    if value > max_val:
        max_val = value
        max_idx = i

visited = [False for _ in range(n)]
dist = [0 for _ in range(n)]
order = 0
visited[max_idx] = True
dfs(max_idx)

for i, value in enumerate(dist):
    if value > max_val:
        max_val = value
        max_idx = i

print(max_val)

File: 240528/트리의 지름/diameter-of-tree.cpp
#include <iostream>
#include <queue>
#include <tuple>
#include <algorithm>
#include <vector>

using namespace std;

int n;
vector<pair<int, int>> G[100000];
bool visited[100000];
int dist[100000];

void    bfs(int s)
{
    queue <pair<int, int>> q;
    visited[s] = true;
    dist[s] = 0;
    q.push(make_pair(s, 0));
    while(!q.empty())
    {
        int x, w;
        tie(x, w) = q.front();
        q.pop();
        for (int i = 0; i < G[x].size(); i++)
        {
            int to_idx, to_dist;
            tie(to_idx, to_dist) = G[x][i];
            if (!visited[to_idx])
            {
                visited[to_idx] = true;
                dist[to_idx] = w + to_dist;
                q.push(make_pair(to_idx, dist[to_idx]));
            }
        }
    }
}

int main() {
    cin >> n;
    for(int i = 0; i < n - 1; i++)
    {
        int s, e, w;
        cin >> s >> e >> w;
        G[s - 1].push_back(make_pair(e - 1, w));
        G[e - 1].push_back(make_pair(s - 1, w));
    }

    bfs(0);    
    int max_val = -1, max_idx = -1;
    for (int i = 0; i < n; i++)
    {
        if (dist[i] > max_val)
        {
            max_val = dist[i];
            max_idx = i;
        }
    }

    for (int i = 0; i < n; i++)
    {
        dist[i] = 0;
        visited[i] = false;
    }

    bfs(max_idx);
    max_val = -1;
    max_idx = -1;
    for (int i = 0; i < n; i++)
    {
        if (dist[i] > max_val)
        {
            max_val = dist[i];
            max_idx = i;
        }
    }

    cout << max_val;
    return 0;
}